### PR TITLE
#7952 New Stencil Buffer Management

### PR DESCRIPTION
**Pull request by Mugen87**
This PR enables three.js to manage the stencil buffer via material properties. 

As mentioned in #7952, i had some problems to implement the masking in context of post-processing with the new material properties. But now i found a way to achieve this in a meaningful way.
1. If a scene wants to use post-processing with masking, users only have to do one additional step with the new approach. They must provide a masking material like in `webgl_postprocessing_masking.html`. That means a material with `stencilWrite` and `stencilTest` set to `true`.
2. The actual post-processing pass chain does not change. If you want to start masking, you need an instance of `THREE.MaskPass`. If you want to stop it, you must use `THREE.ClearMaskPass`.
3. If `THREE.EffectComposer` detects a `THREE.MaskPass`, it activates masking and informs each upcoming pass about the new state (see parameter `maskActive`of render method).
4. A pass evaluates this parameter and sets its internal `quad.material.stencilTest` to the respective value (this will enable/disable the stencil test ONLY for this pass).

With this new approach, direct manipulation of the WebGL State is no longer required. `THREE.WebGLRenderer` handles everything over material properties.

### CODE REVIEW

File: examples/js/postprocessing/GlitchPass.js [hunk 1]
```diff
@@ -72,8 +72,10 @@ THREE.GlitchPass.prototype = {
 
 		}
 		this.curF ++;
-		
+
 		this.quad.material = this.material;
+    this.quad.material.stencilTest = maskActive;
```
Comment (mrdoob) [inline examples/js/postprocessing/GlitchPass.js]: Minor thing: Wrong indent.

**Comment by mrdoob** [comment]
Looks good!

File: examples/js/postprocessing/GlitchPass.js (hunk 1 above)
Comment (Mugen87) [inline examples/js/postprocessing/GlitchPass.js]: I'll fix this...

**Comment by Mugen87** [comment]
I think i should check my formatting options in my editor (Atom). In rare cases, i have the problem that my code looks accurate on my Mac, but on github i see things like the mentioned wrong indent.

**Comment by WestLangley** [comment]
http://ci.threejs.org/api/pullrequests/8006/examples/webgl_postprocessing_masking.html is rendering with black material for me. OS X 10.11.2  Chrome 47.0.2526.111, Safari 9.0.2.

**Comment by Mugen87** [comment]
It looks like the textures are not loaded correctly. When i look in my network module (Chrome Dev Tools), the status of the textures is (pending). Maybe a server or network problem?

**Comment by Mugen87** [comment]
This is the direct link to one of those textures. Always getting status "pending"...

http://ci.threejs.org/api/pullrequests/8006/examples/textures/758px-Canestra_di_frutta_(Caravaggio).jpg

**Comment by mrdoob** [comment]
Yeah, works fine locally.

**Comment by mrdoob** [comment]
Thanks!

**Comment by WestLangley** [comment]
@Mugen87 Would you mind modiying the masking example so it just specifies a mask and then renders a simple scene? That would be a more typical use-case, IMO.

	composer.addPass( maskPass );
	composer.addPass( renderPass );

**Comment by Mugen87** [comment]
The minimal setup for such a use case is this:

```javascript

composer.addPass( renderPass );
composer.addPass( maskPass );
composer.addPass( copyPass );
composer.addPass( clearMaskPass );
composer.addPass( outputPass );

```
Masking only works if a pass uses a quad for rendering. So it's necessary to do the following steps:

* render the actual scene to the read-buffer( `renderPass`)
* enable masking (`maskPass`)
* copy the read-buffer to write-buffer  with masking (`copyPass`). This pass also swaps internally read/write buffer.
* disable masking (`clearMaskPass`)
* write the read-buffer to the default FBO (`outputPass`)

The result looks like [this](http://yume.human-interactive.org/test/postprocessing/webgl_postprocessing_masking.html).

**Comment by Mugen87** [comment]
Oh, i just realized that such basic masking does not require post processing. Just have a look at this [example](http://yume.human-interactive.org/test/postprocessing/masking_simple.html). Same effect, but without post processing. The method `updateMasking` does the whole magic.

https://github.com/mrdoob/three.js/pull/7836#issuecomment-166947922

>We could probably add a portal kind of example.

Maybe it's better to add a portal example instead of modifying the existing one. Masking/Portal or Window effect are basically the same things in context of the stencil buffer....

**Comment by mrdoob** [comment]
That's a bit advanced, but would be nice to add that as an example too!

**Comment by WestLangley** [comment]
@Mugen87 I am having difficulty getting a previous masking example to work using your new system, which is why I am asking you these questions...

>render the actual scene to the read-buffer( renderPass )
>enable masking ( maskPass )

It seems to me that one would naturally set the mask first, and then render the scene.

Can your new system be made to work that way? I feel your approach is too limiting.

**Comment by Mugen87** [comment]
> It seems to me that one would naturally set the mask first, and then render the scene.

I see what you mean. The problem is, that the stencil test is now controlled by material properties (`stencilTest`). I see currently just one (dirty) way to achieve an immediate masking in `THREE.RenderPass`:

* check parameter `maskActive`
* if set to true, traverse the scene and set the `stencilTest` to `true` for all 3d objects
* call the render method of the renderer
* traverse the scene again and undo all changes

I would never implement something like this, but it illustrate the problem to fulfill your requirement. Maybe we need a little enhancement so the renderer can override the material property `stencilTest` (something like overrideMaterial in `THREE.Scene`). Have to think about this....

**Comment by Mugen87** [comment]
@WestLangley What do you think about this [example](http://yume.human-interactive.org/test/masking/webgl_postprocessing_masking2.html)? It uses a slightly adjusted implementation.

Post-processing chain:

```javascript
composer.addPass( clearPass );
composer.addPass( maskPass );
composer.addPass( renderPass );
composer.addPass( clearMaskPass );
composer.addPass( outputPass );
```

Please have a look at the updated `render` method of `THREE.RenderPass`. If masking is active, it uses the new method `forceStencilTest` of `THREE.WebGLState` to force the stencil test for all 3D objects in the scene. Is this too...hacky?

**Comment by MasterJames** [comment]
It's flickering wildly on my Samsung Note3 Android mobile phone.

**Pull request merged**